### PR TITLE
Add ModelCheckpoint.to_yaml method

### LINKED ISSUE #3047: ModelCheckpoint instance to JSON
## 🚀 Feature

Have a `to_json(path)` function that writes the `best_k_models` to a json file in `path`.

### Motivation

Why having a JSON is good? 
Then you don't need to load the checkpoints to know the metrics and things like a posteriori stochastic weight averaging gets  easier. 

It used to be easier in 0.6, but loss values are now tensors, so it's not directly serializable.
This is currently needed: 
```python
best_k = {k: v.item() for k, v in checkpoint.best_k_models.items()}
with open(json_path, "w") as f:
    json.dump(best_k, f, indent=0)
```

Instead, we could have 
```python
checkpoint.to_json(json_path)
```

I'm ready to make a PR for it. Just tell me if this would be useful for you.

### PR DESCRIPTION
**Pull request by mpariente**
## What does this PR do

Add simple `to_json()` method to `ModelCheckpoint`. 
In the current code, it requires three lines, and it could be automated by lightning.
Fixes #3047 (disclaimer: my issue with no answer yet)

# Before submitting

- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by justusschock** [comment]
Actually I'd prefer not to use json for this. If we really want this (which still needs to be considered), I'd prefer to go with YAML instead of JSON here, since we use yaml for most other human readable serialization and IMO should not mix file formats to much.

**Comment by rohitgr7** [comment]
why not both or maybe more? `to_some_format(format='YAML' or 'JSON' or ...)`?
only if we really want this (which still needs to be considered)

**Comment by mpariente** [comment]
> If we really want this (which still needs to be considered)

I can make the required changes, but I'd like to know if you want it or not beforehand.

**Comment by Borda** [comment]
personally I would prefer YAML as it is readable and it can be used as an architecture draft, but we can have both...
any other thoughts on the API side? @PyTorchLightning/core-contributors

**Comment by justusschock** [comment]
I would strongly recommend not to do both, since this is really not necessary here. It's just so that you know which checkpoints to use and to consider. Doing both is a typical overkill and we still have to keep the effort to maintain in mind (probably not much here, but it increases with every little bit).

**Comment by mpariente** [comment]
YAML only would be ok. It's just to have a one-liner to dump the saved models history. 

I think there is a point to this feature because having this file is useful after training, for stochastic weight averaging or ensemble methods etc.. And if we don't dump this file, we have to open the checkpoint to have it's validation performance, which is not efficient.

**Comment by rohitgr7** [comment]
yeah, YAML sounds good. Let's do YAML then :).

**Comment by Borda** [comment]
@mpariente shall we close this one and make another for YAML?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
 
 import os
 import re
+import json
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:25]:
```suggestion
import yaml
```

File: tests/callbacks/test_model_checkpoint.py [hunk 2]
```diff
@@ -34,6 +35,23 @@ def test_model_checkpoint_with_non_string_input(tmpdir, save_top_k):
...
+
+    checkpoint = ModelCheckpoint(filepath=None, save_top_k=save_top_k)
+
+    trainer = Trainer(default_root_dir=tmpdir, checkpoint_callback=checkpoint, overfit_batches=0.20, max_epochs=2)
+    trainer.fit(model)
+
+    checkpoint.to_json('./best_k_models.json')
+    d = json.load(open('./best_k_models.json', 'r'))
+    best_k = {k: torch.Tensor(v) for k, v in d.items()}
+    torch.testing.assert_allclose(best_k, checkpoint.best_k_models)
```
Comment (Borda) [inline tests/callbacks/test_model_checkpoint.py:52]:
```suggestion
    checkpoint.to_yaml('./best_k_models.yaml')
    d = yaml.full_load(open('./best_k_models.yaml', 'r'))
    best_k = {k: torch.Tensor(v) for k, v in d.items()}
    torch.testing.assert_allclose(best_k, checkpoint.best_k_models)
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -466,3 +467,11 @@ def on_save_checkpoint(
...
         self.best_model_score = checkpointed_state["best_model_score"]
         self.best_model_path = checkpointed_state["best_model_path"]
+
+    def to_json(self, save_path=None):
+        """ Saves `{'checkpoint_name': score}` dict as a JSON file."""
+        best_k = {k: v.item() for k, v in self.best_k_models.items()}
+        if save_path is None:
+            save_path = os.path.join(self.dirpath, "best_k_models.json")
+        with open(save_path, "w") as f:
+            json.dump(best_k, f, indent=0)
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:477]:
```suggestion
    def to_yaml(self, save_path=None):
        """ Saves `{'checkpoint_name': score}` dict as a YAML file."""
        best_k = {k: v.item() for k, v in self.best_k_models.items()}
        if save_path is None:
            save_path = os.path.join(self.dirpath, "best_k_models.yaml")
        with open(save_path, "w") as fp:
            yaml.dump(best_k, fp)
```

File: tests/callbacks/test_model_checkpoint.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import os
+import json
```
Comment (Borda) [inline tests/callbacks/test_model_checkpoint.py:2]:
```suggestion
import yaml
```

**Review by Borda — APPROVED** [review]
LGTM

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -466,3 +467,11 @@ def on_save_checkpoint(
     def on_load_checkpoint(self, checkpointed_state: Dict[str, Any]):
         self.best_model_score = checkpointed_state["best_model_score"]
         self.best_model_path = checkpointed_state["best_model_path"]
+
+    def to_yaml(self, save_path=None):
```
Comment (justusschock) [inline pytorch_lightning/callbacks/model_checkpoint.py:471]:
```suggestion
    def to_yaml(self, save_path: Optional[Union[str, Path]]=None):
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 6]
```diff
@@ -22,6 +22,7 @@
 
 import os
 import re
+import yaml
 from copy import deepcopy
 from typing import Any, Dict, Optional
```
Comment (justusschock) [inline pytorch_lightning/callbacks/model_checkpoint.py:27]:
```suggestion
from typing import Any, Dict, Optional, Union
from pathlib import Path
```

**Review by justusschock — APPROVED** [review]

**Comment by mpariente** [comment]
Sorry, I got absorbed by something else and forgot the PR. 
Thanks for pushing this through @Borda !

**Comment by Borda** [comment]
@mpariente seems that with the update to YAML the test is not correct, mind have look at it...

**Comment by mpariente** [comment]
This week end, I won't have time. 
But I'll have a look on Monday.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 7]
```diff
@@ -466,3 +468,11 @@ def on_save_checkpoint(
...
         self.best_model_score = checkpointed_state["best_model_score"]
         self.best_model_path = checkpointed_state["best_model_path"]
+
+    def to_yaml(self, save_path: Optional[Union[str, Path]]=None):
+        """ Saves `{'checkpoint_name': score}` dict as a YAML file."""
+        best_k = {k: v.item() for k, v in self.best_k_models.items()}
+        if save_path is None:
+            save_path = os.path.join(self.dirpath, "best_k_models.yaml")
+        with open(save_path, "w") as fp:
+            yaml.dump(best_k, fp)
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:478]:
```suggestion
    def to_yaml(self, filepath: Optional[Union[str, Path]] = None):
        """
        Saves the `best_k_models` dict containing the checkpoint
        paths with the corresponding scores to a YAML file.
        """
        best_k = {k: v.item() for k, v in self.best_k_models.items()}
        if filepath is None:
            filepath = os.path.join(self.dirpath, "best_k_models.yaml")
        with open(filepath, "w") as fp:
            yaml.dump(best_k, fp)
```

**Review by rohitgr7 — APPROVED** [review]

**Comment by Borda** [comment]
> This week end, I won't have time.
> But I'll have a look on Monday.

It seems you did it, thank you :]

File: tests/callbacks/test_model_checkpoint.py [hunk 8]
```diff
@@ -36,6 +37,23 @@ def test_model_checkpoint_with_non_string_input(tmpdir, save_top_k):
...
+    tutils.reset_seed()
+    model = EvalModelTemplate()
+
+    checkpoint = ModelCheckpoint(filepath=None, save_top_k=save_top_k)
+
+    trainer = Trainer(default_root_dir=tmpdir, checkpoint_callback=checkpoint, overfit_batches=0.20, max_epochs=2)
+    trainer.fit(model)
+
+    checkpoint.to_yaml('./best_k_models.yaml')
+    d = yaml.full_load(open('./best_k_models.yaml', 'r'))
```
Comment (Borda) [inline tests/callbacks/test_model_checkpoint.py:52]:
```suggestion
    checkpoint.to_yaml(os.path.join(tmpdir, 'best_k_models.yaml'))
    d = yaml.full_load(openos.path.join(tmpdir, 'best_k_models.yaml'), 'r'))
```

**Comment by mpariente** [comment]
Yeah, finally ^^ 
What happening with the tests, you rebased master? The failures are not related right?

File: tests/callbacks/test_model_checkpoint.py [hunk 9]
```diff
@@ -36,6 +37,24 @@ def test_model_checkpoint_with_non_string_input(tmpdir, save_top_k):
     )
 
 
+@pytest.mark.parametrize('save_top_k', [-1, 0, 1, 2])
+def test_model_checkpoint_to_yaml(tmpdir, save_top_k):
+    """ Test that None in checkpoint callback is valid and that chkp_path is set correctly """
+    tutils.reset_seed()
+    model = EvalModelTemplate()
+
+    checkpoint = ModelCheckpoint(filepath=None, save_top_k=save_top_k)
```
Comment (rohitgr7) [inline tests/callbacks/test_model_checkpoint.py:46]:
```suggestion
    checkpoint = ModelCheckpoint(filepath=tmpdir, monitor='val_loss', save_top_k=save_top_k)
```

**Comment by rohitgr7** [comment]
good to go now @Borda

**Comment by Borda** [comment]
Well, just forget to add changelog =)
